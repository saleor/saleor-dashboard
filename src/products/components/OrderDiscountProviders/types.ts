import { MoneyFragment, TaxedMoneyFragment } from "@saleor/graphql";
import { OrderDiscountCommonInput } from "@saleor/orders/components/OrderDiscountCommonModal/types";

import { OrderLineDiscountContextConsumerProps } from "./OrderLineDiscountProvider";

export interface OrderDiscountData extends OrderDiscountCommonInput {
  amount: MoneyFragment;
}

export type GetOrderLineDiscountContextConsumerProps = (
  orderLineId: string,
) => OrderLineDiscountContextConsumerProps;

export interface OrderLineDiscountData extends OrderDiscountCommonInput {
  moneyValue: MoneyFragment;
  undiscountedPrice: TaxedMoneyFragment;
}

export interface OrderDiscountConsumerCommonProps {
  openDialog: () => void;
  closeDialog: () => void;
  isDialogOpen: boolean;
  undiscountedPrice: MoneyFragment;
  discountedPrice: MoneyFragment;
}

export interface OrderLineDiscountConsumerProps {
  children: (values: OrderLineDiscountContextConsumerProps) => React.ReactNode;
  orderLineId: string;
}
