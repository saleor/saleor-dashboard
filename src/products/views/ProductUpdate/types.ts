export interface ChannelWithVariantData {
  selectedVariantsIds: string[];
  variantsIdsToRemove: string[];
  variantsIdsToAdd: string[];
}

export type ChannelVariantListing = Record<string, string[]>;
export type ChannelsWithVariantsData = Record<string, ChannelWithVariantData>;

export interface UseChannelsWithProductVariants {
  addVariantToChannel: (channelId: string, variantId: string) => void;
  removeVariantFromChannel: (channelId: string, variantId: string) => void;
  channelsWithVariantsData: ChannelsWithVariantsData;
  toggleAllChannels: () => void;
  toggleAllChannelVariants: (channelId: string) => void;
  channelVariantListing: ChannelVariantListing;
  setChannelVariantListing: (listings: ChannelVariantListing) => void;
  reset: () => void;
}
