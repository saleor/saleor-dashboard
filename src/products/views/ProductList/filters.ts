import isArray from "lodash-es/isArray";

import { maybe, findValueInEnum } from "@saleor/misc";
import {
  ProductFilterKeys,
  ProductListFilterOpts,
  ProductStatus
} from "@saleor/products/components/ProductListPage";
import { UseSearchResult } from "@saleor/hooks/makeSearch";
import {
  SearchCategories,
  SearchCategoriesVariables
} from "@saleor/searches/types/SearchCategories";
import {
  InitialProductFilterData_categories_edges_node,
  InitialProductFilterData_collections_edges_node,
  InitialProductFilterData_productTypes_edges_node,
  InitialProductFilterData_attributes_edges_node
} from "@saleor/products/types/InitialProductFilterData";
import {
  SearchCollections,
  SearchCollectionsVariables
} from "@saleor/searches/types/SearchCollections";
import {
  SearchProductTypes,
  SearchProductTypesVariables
} from "@saleor/searches/types/SearchProductTypes";
import { IFilterElement } from "../../../components/Filter";
import {
  ProductFilterInput,
  StockAvailability
} from "../../../types/globalTypes";
import {
  createFilterTabUtils,
  createFilterUtils,
  getGteLteVariables,
  getMinMaxQueryParam,
  getSingleEnumValueQueryParam,
  dedupeFilter,
  getMultipleValueQueryParam
} from "../../../utils/filters";
import {
  ProductListUrlFilters,
  ProductListUrlFiltersEnum,
  ProductListUrlQueryParams,
  ProductListUrlFiltersWithMultipleValues
} from "../../urls";

export const PRODUCT_FILTERS_KEY = "productFilters";

export function getFilterOpts(
  params: ProductListUrlFilters,
  attributes: InitialProductFilterData_attributes_edges_node[],
  categories: {
    initial: InitialProductFilterData_categories_edges_node[];
    search: UseSearchResult<SearchCategories, SearchCategoriesVariables>;
  },
  collections: {
    initial: InitialProductFilterData_collections_edges_node[];
    search: UseSearchResult<SearchCollections, SearchCollectionsVariables>;
  },
  productTypes: {
    initial: InitialProductFilterData_productTypes_edges_node[];
    search: UseSearchResult<SearchProductTypes, SearchProductTypesVariables>;
  }
): ProductListFilterOpts {
  return {
    attributes: attributes
      .sort((a, b) => (a.name > b.name ? 1 : -1))
      .map(attr => ({
        active: maybe(() => params.attributes[attr.slug].length > 0, false),
        choices: attr.values.map(val => ({
          label: val.name,
          value: val.slug
        })),
        name: attr.name,
        slug: attr.slug,
        value:
          !!params.attributes && params.attributes[attr.slug]
            ? params.attributes[attr.slug]
            : []
      })),
    categories: {
      active: !!params.categories,
      choices: maybe(
        () =>
          categories.search.result.data.search.edges.map(edge => ({
            label: edge.node.name,
            value: edge.node.id
          })),
        []
      ),
      displayValues: !!params.categories
        ? maybe(
            () =>
              categories.initial.map(category => ({
                label: category.name,
                value: category.id
              })),
            []
          )
        : [],
      hasMore: maybe(
        () => categories.search.result.data.search.pageInfo.hasNextPage,
        false
      ),
      initialSearch: "",
      loading: categories.search.result.loading,
      onFetchMore: categories.search.loadMore,
      onSearchChange: categories.search.search,
      value: maybe(() => dedupeFilter(params.categories), [])
    },
    collections: {
      active: !!params.collections,
      choices: maybe(
        () =>
          collections.search.result.data.search.edges.map(edge => ({
            label: edge.node.name,
            value: edge.node.id
          })),
        []
      ),
      displayValues: !!params.collections
        ? maybe(
            () =>
              collections.initial.map(category => ({
                label: category.name,
                value: category.id
              })),
            []
          )
        : [],
      hasMore: maybe(
        () => collections.search.result.data.search.pageInfo.hasNextPage,
        false
      ),
      initialSearch: "",
      loading: collections.search.result.loading,
      onFetchMore: collections.search.loadMore,
      onSearchChange: collections.search.search,
      value: maybe(() => dedupeFilter(params.collections), [])
    },
    price: {
      active: maybe(
        () =>
          [params.priceFrom, params.priceTo].some(field => field !== undefined),
        false
      ),
      value: {
        max: maybe(() => params.priceTo, "0"),
        min: maybe(() => params.priceFrom, "0")
      }
    },
    productType: {
      active: !!params.productTypes,
      choices: maybe(
        () =>
          productTypes.search.result.data.search.edges.map(edge => ({
            label: edge.node.name,
            value: edge.node.id
          })),
        []
      ),
      displayValues: !!params.productTypes
        ? maybe(
            () =>
              productTypes.initial.map(productType => ({
                label: productType.name,
                value: productType.id
              })),
            []
          )
        : [],
      hasMore: maybe(
        () => productTypes.search.result.data.search.pageInfo.hasNextPage,
        false
      ),
      initialSearch: "",
      loading: productTypes.search.result.loading,
      onFetchMore: productTypes.search.loadMore,
      onSearchChange: productTypes.search.search,
      value: maybe(() => dedupeFilter(params.productTypes), [])
    },
    status: {
      active: maybe(() => params.status !== undefined, false),
      value: maybe(() => findValueInEnum(params.status, ProductStatus))
    },
    stockStatus: {
      active: maybe(() => params.stockStatus !== undefined, false),
      value: maybe(() => findValueInEnum(params.stockStatus, StockAvailability))
    }
  };
}

export function getFilterVariables(
  params: ProductListUrlFilters
): ProductFilterInput {
  return {
    attributes: !!params.attributes
      ? Object.keys(params.attributes).map(key => ({
          slug: key,
          // It is possible for qs to parse values not as string[] but string
          values: isArray(params.attributes[key])
            ? params.attributes[key]
            : (([params.attributes[key]] as unknown) as string[])
        }))
      : null,
    categories: params.categories !== undefined ? params.categories : null,
    collections: params.collections !== undefined ? params.collections : null,
    isPublished:
      params.status !== undefined
        ? params.status === ProductStatus.PUBLISHED
        : null,
    price: getGteLteVariables({
      gte: parseFloat(params.priceFrom),
      lte: parseFloat(params.priceTo)
    }),
    productTypes:
      params.productTypes !== undefined ? params.productTypes : null,
    search: params.query,
    stockAvailability:
      params.stockStatus !== undefined
        ? findValueInEnum(params.stockStatus, StockAvailability)
        : null
  };
}

export function getFilterQueryParam(
  filter: IFilterElement<ProductFilterKeys>,
  params: ProductListUrlFilters
): ProductListUrlFilters {
  const { active, group, name, value } = filter;

  if (!!group) {
    const rest = params && params[group] ? params[group] : undefined;

    return {
      [group]: active
        ? {
            ...(rest === undefined ? {} : rest),
            [name]: value
          }
        : rest
    };
  }

  switch (name) {
    case ProductFilterKeys.categories:
      return getMultipleValueQueryParam(
        filter,
        ProductListUrlFiltersWithMultipleValues.categories
      );

    case ProductFilterKeys.collections:
      return getMultipleValueQueryParam(
        filter,
        ProductListUrlFiltersWithMultipleValues.collections
      );

    case ProductFilterKeys.price:
      return getMinMaxQueryParam(
        filter,
        ProductListUrlFiltersEnum.priceFrom,
        ProductListUrlFiltersEnum.priceTo
      );

    case ProductFilterKeys.productType:
      return getMultipleValueQueryParam(
        filter,
        ProductListUrlFiltersWithMultipleValues.productTypes
      );

    case ProductFilterKeys.status:
      return getSingleEnumValueQueryParam(
        filter,
        ProductListUrlFiltersEnum.status,
        ProductStatus
      );

    case ProductFilterKeys.stock:
      return getSingleEnumValueQueryParam(
        filter,
        ProductListUrlFiltersEnum.stockStatus,
        StockAvailability
      );
  }
}

export const {
  deleteFilterTab,
  getFilterTabs,
  saveFilterTab
} = createFilterTabUtils<ProductListUrlFilters>(PRODUCT_FILTERS_KEY);

export const { areFiltersApplied, getActiveFilters } = createFilterUtils<
  ProductListUrlQueryParams,
  ProductListUrlFilters
>(ProductListUrlFiltersEnum);
