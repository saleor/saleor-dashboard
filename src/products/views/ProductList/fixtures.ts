import { attributes } from "@saleor/attributes/fixtures";
import { categories } from "@saleor/categories/fixtures";
import { collections } from "@saleor/collections/fixtures";
import { fetchMoreProps, searchPageProps } from "@saleor/fixtures";
import {
  ProductListFilterOpts,
  ProductStatus
} from "@saleor/products/components/ProductListPage";
import { productTypes } from "@saleor/productTypes/fixtures";
import { StockAvailability } from "@saleor/types/globalTypes";

export const productListFilterOpts: ProductListFilterOpts = {
  attributes: attributes.map(attr => ({
    active: false,
    choices: attr.values.map(val => ({
      label: val.name,
      value: val.slug
    })),
    name: attr.name,
    slug: attr.slug,
    value: [attr.values[0].slug, attr.values.length > 2 && attr.values[2].slug]
  })),
  categories: {
    ...fetchMoreProps,
    ...searchPageProps,
    active: false,
    choices: categories.slice(5).map(category => ({
      label: category.name,
      value: category.id
    })),
    displayValues: [
      {
        label: categories[5].name,
        value: categories[5].id
      }
    ],
    value: [categories[5].id]
  },
  collections: {
    ...fetchMoreProps,
    ...searchPageProps,
    active: false,
    choices: collections.slice(5).map(category => ({
      label: category.name,
      value: category.id
    })),
    displayValues: [
      {
        label: collections[5].name,
        value: collections[5].id
      }
    ],
    value: [collections[5].id]
  },
  price: {
    active: false,
    value: {
      max: "20",
      min: "10"
    }
  },
  productType: {
    ...fetchMoreProps,
    ...searchPageProps,
    active: false,
    choices: productTypes.slice(3).map(category => ({
      label: category.name,
      value: category.id
    })),
    displayValues: [
      {
        label: productTypes[3].name,
        value: productTypes[3].id
      }
    ],
    value: [productTypes[4].id]
  },
  status: {
    active: false,
    value: ProductStatus.PUBLISHED
  },
  stockStatus: {
    active: false,
    value: StockAvailability.IN_STOCK
  }
};
