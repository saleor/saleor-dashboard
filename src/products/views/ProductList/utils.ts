import { ProductTypeKindEnum } from "@saleor/graphql";
import { isInEnum } from "@saleor/misc";
import { IntlShape } from "react-intl";

import { productKindMessages as messages } from "./messages";

interface ProductKindChoice {
  label: string;
  value: ProductTypeKindEnum;
}

export const getAvailableProductKinds = (): ProductKindChoice[] =>
  Object.keys(ProductTypeKindEnum).map(kind => ({
    label: kind,
    value: kind as ProductTypeKindEnum,
  }));

export const getProductKindOpts = (
  availableProducts: ProductKindChoice[],
  intl: IntlShape,
): ProductKindChoice[] =>
  availableProducts.map(kind => {
    switch (kind.value) {
      case ProductTypeKindEnum.GIFT_CARD:
        return {
          ...kind,
          label: intl.formatMessage(messages.giftCardLabel),
        };
      case ProductTypeKindEnum.NORMAL:
        return {
          ...kind,
          label: intl.formatMessage(messages.normalLabel),
        };
    }
  });

export const getProductGiftCardFilterParam = (productKind?: string) => {
  if (
    productKind === undefined ||
    !isInEnum(productKind, ProductTypeKindEnum)
  ) {
    return null;
  }

  return productKind === ProductTypeKindEnum.GIFT_CARD;
};
