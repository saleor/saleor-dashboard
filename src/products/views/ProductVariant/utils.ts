import { WarehouseListQuery } from "@dashboard/graphql";

export const mergeWarehousesQuery = (
  previousResult: WarehouseListQuery,
  {
    fetchMoreResult,
  }: {
    fetchMoreResult?: WarehouseListQuery;
  },
): WarehouseListQuery => {
  if (!fetchMoreResult) {
    return previousResult;
  }

  const previousEdges = previousResult?.warehouses?.edges ?? [];
  const fetchMoreEdges = fetchMoreResult?.warehouses?.edges ?? [];

  if (fetchMoreResult?.warehouses?.edges) {
    fetchMoreResult.warehouses.edges = [...previousEdges, ...fetchMoreEdges];

    return { ...fetchMoreResult };
  }

  return previousResult;
};
