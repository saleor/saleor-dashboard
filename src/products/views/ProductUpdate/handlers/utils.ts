import {
  ChannelData,
  createSortedChannelsDataFromProduct,
} from "@saleor/channels/utils";
import {
  ProductChannelListingAddInput,
  ProductDetailsVariantFragment,
  ProductFragment,
  ProductUpdateMutationVariables,
  SimpleProductUpdateMutation,
} from "@saleor/graphql";
import { weight } from "@saleor/misc";
import { getById } from "@saleor/orders/components/OrderReturnPage/utils";
import { ProductUpdatePageSubmitData } from "@saleor/products/components/ProductUpdatePage";
import { ProductUpdateSubmitData } from "@saleor/products/components/ProductUpdatePage/form";
import { mapFormsetStockToStockInput } from "@saleor/products/utils/data";
import { getAvailabilityVariables } from "@saleor/products/utils/handlers";
import { arrayDiff } from "@saleor/utils/arrays";
import isEqual from "lodash/isEqual";

import { ChannelsWithVariantsData, ChannelWithVariantData } from "../types";
import { getParsedChannelsWithVariantsDataFromChannels } from "../utils";

export const getSimpleProductVariables = (
  productVariables: ProductUpdateMutationVariables,
  data: ProductUpdatePageSubmitData,
  productId: string,
) => ({
  ...productVariables,
  addStocks: data.addStocks.map(mapFormsetStockToStockInput),
  deleteStocks: data.removeStocks,
  input: {
    ...productVariables.input,
    weight: weight(data.weight),
  },
  productVariantId: productId,
  productVariantInput: {
    sku: data.sku,
    trackInventory: data.trackInventory,
    preorder: data.isPreorder
      ? {
          globalThreshold: data.globalThreshold
            ? parseInt(data.globalThreshold, 10)
            : null,
          endDate: data.preorderEndDateTime,
        }
      : undefined,
  },
  updateStocks: data.updateStocks.map(mapFormsetStockToStockInput),
});

export const getSimpleProductErrors = (data: SimpleProductUpdateMutation) => [
  ...data.productUpdate.errors,
  ...data.productVariantStocksCreate.errors,
  ...data.productVariantStocksDelete.errors,
  ...data.productVariantStocksUpdate.errors,
];

export const getChannelListingBaseInputData = ({
  id: channelId,
  isPublished,
  publicationDate,
  isAvailableForPurchase,
  availableForPurchase,
  visibleInListings,
}: ChannelData) => ({
  channelId,
  isPublished,
  publicationDate,
  visibleInListings,
  isAvailableForPurchase,
  availableForPurchaseDate: availableForPurchase,
});

export const getChannelListingUpdateInputFromData = (
  { variantsIdsToAdd, variantsIdsToRemove }: ChannelWithVariantData,
  { selectedVariantsIds: initialSelectedVariantsIds }: ChannelWithVariantData,
  basicChannelData: ChannelData,
) => ({
  ...getChannelListingBaseInputData(basicChannelData),
  addVariants: arrayDiff(initialSelectedVariantsIds, variantsIdsToAdd).added,
  removeVariants: variantsIdsToRemove,
});

const getParsedChannelsData = (
  channelsWithVariants: ChannelsWithVariantsData,
  initialChannelWithVariants: ChannelsWithVariantsData,
  channelsData: ChannelData[],
): ProductChannelListingAddInput[] =>
  channelsData.map(({ id, ...rest }) =>
    getChannelListingUpdateInputFromData(
      channelsWithVariants[id],
      initialChannelWithVariants[id],
      { id, ...rest },
    ),
  );

const shouldRemoveChannel = (allVariants: ProductDetailsVariantFragment[]) => ({
  removeVariants,
}: ProductChannelListingAddInput) =>
  isRemovingAllVariants(allVariants, removeVariants);

const isRemovingAllVariants = (
  allVariants: ProductDetailsVariantFragment[],
  removeVariants: string[],
) => !!removeVariants.length && removeVariants.length === allVariants.length;

const shouldUpdateChannel = (
  initialChannelWithVariantData,
  allVariants: ProductDetailsVariantFragment[],
  allChannels: ChannelData[],
) => ({
  removeVariants,
  addVariants,
  channelId,
  ...rest
}: ProductChannelListingAddInput) => {
  const initialDataInput = getChannelListingUpdateInputFromData(
    initialChannelWithVariantData[channelId],
    initialChannelWithVariantData[channelId],
    allChannels.find(getById(channelId)),
  );

  const hasDataChanged = !isEqual(
    { removeVariants, addVariants, channelId, ...rest },
    initialDataInput,
  );

  const isRemovingChannel = isRemovingAllVariants(allVariants, removeVariants);

  return hasDataChanged && !isRemovingChannel;
};

export const getChannelsVariables = (
  { id, variants }: ProductFragment,
  allChannels: ChannelData[],
  { channelsWithVariants, channelsData }: ProductUpdateSubmitData,
) => {
  const initialChannelWithVariants = getParsedChannelsWithVariantsDataFromChannels(
    channelsData,
  );

  const channelsToBeUpdated = getParsedChannelsData(
    channelsWithVariants,
    initialChannelWithVariants,
    channelsData,
  ).filter(
    shouldUpdateChannel(initialChannelWithVariants, variants, allChannels),
  );

  const channelsIdsToBeRemoved = getParsedChannelsData(
    channelsWithVariants,
    initialChannelWithVariants,
    channelsData,
  )
    .filter(shouldRemoveChannel(variants))
    .map(({ channelId }) => channelId);

  return {
    variables: {
      id,
      input: {
        updateChannels: channelsToBeUpdated,
        removeChannels: channelsIdsToBeRemoved,
      },
    },
  };
};

export const getSimpleChannelsVariables = (
  data: ProductUpdatePageSubmitData,
  product: ProductFragment,
) => {
  const productChannels = createSortedChannelsDataFromProduct(product);
  const existingChannelIDs = productChannels.map(channel => channel.id);
  const modifiedChannelIDs = data.channelListings.map(channel => channel.id);

  const removedChannelIDs = existingChannelIDs.filter(
    x => !modifiedChannelIDs.includes(x),
  );

  return {
    variables: {
      id: product.id,
      input: {
        updateChannels: getAvailabilityVariables(data.channelListings),
        removeChannels: removedChannelIDs,
      },
    },
  };
};

export const getVariantChannelsInput = ({
  channelListings,
}: ProductUpdatePageSubmitData) =>
  channelListings.map(listing => ({
    channelId: listing.id,
    costPrice: listing.costPrice || null,
    price: listing.price,
    preorderThreshold: listing.preorderThreshold,
  }));
