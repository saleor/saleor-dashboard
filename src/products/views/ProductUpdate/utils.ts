import { ChannelData } from "@saleor/channels/utils";
import every from "lodash/every";
import reduce from "lodash/reduce";

import { initialChannelWithVariantData } from "./consts";
import {
  ChannelsWithVariantsData,
  ChannelVariantListing,
  ChannelWithVariantData,
} from "./types";

export function createFromChannels<T>(
  channels: ChannelData[],
  cb: (channel: ChannelData) => T,
): Record<string, T> {
  return channels?.reduce(
    (result: Record<string, T>, channel) => ({
      ...result,
      [channel.id]: cb(channel),
    }),
    {},
  );
}

export function createUpdatedChannels(
  channels: ChannelData[],
  listing: ChannelVariantListing,
): ChannelData[] {
  return reduce(
    listing,
    (acc, variantsIds, channelId) => [
      ...acc,
      {
        id: channelId,
        name: channels.find(channel => channel.id === channelId).name,
        variantsIds,
      } as ChannelData,
    ],
    [],
  );
}

export const getParsedChannelsWithVariantsDataFromChannels = (
  channels: ChannelData[],
): ChannelsWithVariantsData =>
  createFromChannels(
    channels,
    ({ variantsIds }) =>
      ({
        ...initialChannelWithVariantData,
        selectedVariantsIds: variantsIds,
      } as ChannelWithVariantData),
  );

export const getChannelVariantToggleData = (
  variants: string[],
  isSelected: boolean,
): string[] => (isSelected ? [] : variants);

export const areAllVariantsAtAllChannelsSelected = (
  variants: string[] = [],
  channelsWithVariants: ChannelVariantListing = {},
) =>
  every(channelsWithVariants, channelWithVariantsData =>
    areAllChannelVariantsSelected(variants, channelWithVariantsData),
  );

export const areAllChannelVariantsSelected = (
  variants: string[] = [],
  selectedVariants: string[],
) => selectedVariants.length === variants.length;

export const areAnyChannelVariantsSelected = (
  channelsWithVariantsData: ChannelWithVariantData,
) => channelsWithVariantsData?.selectedVariantsIds.length > 0;

export const getTotalSelectedChannelsCount = (
  channelsWithVariantsData: ChannelsWithVariantsData,
) =>
  Object.values(channelsWithVariantsData).filter(
    channel => channel.selectedVariantsIds.length > 0,
  ).length;

export const addAllVariantsToAllChannels = (
  listings: ChannelVariantListing,
  variants: string[],
): ChannelVariantListing => {
  const areAllChannelsSelected = areAllVariantsAtAllChannelsSelected(
    variants,
    listings,
  );

  const updatedListing = reduce(
    listings,
    (result: ChannelVariantListing, _, channelId) => ({
      ...result,
      [channelId]: getChannelVariantToggleData(
        variants,
        areAllChannelsSelected,
      ),
    }),
    {},
  );

  return updatedListing;
};

export const channelVariantListingDiffToDict = (
  listing: ChannelsWithVariantsData,
): ChannelVariantListing =>
  reduce(
    listing,
    (
      listingDict: ChannelVariantListing,
      { selectedVariantsIds },
      channelId,
    ) => ({
      ...listingDict,
      [channelId]: selectedVariantsIds,
    }),
    {},
  );
