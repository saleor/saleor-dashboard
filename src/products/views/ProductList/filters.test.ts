import { createFilterStructure } from "@saleor/products/components/ProductListPage";
import { ProductListUrlFilters } from "@saleor/products/urls";
import { StockAvailability } from "@saleor/types/globalTypes";
import { getFilterQueryParams } from "@saleor/utils/filters";
import { getExistingKeys, setFilterOptsStatus } from "@test/filters";
import { config } from "@test/intl";
import { stringify as stringifyQs } from "qs";
import { createIntl } from "react-intl";

import { getFilterQueryParam, getFilterVariables } from "./filters";
import { productListFilterOpts } from "./fixtures";

describe("Filtering query params", () => {
  it("should be empty object if no params given", () => {
    const params: ProductListUrlFilters = {};
    const filterVariables = getFilterVariables(params);

    expect(getExistingKeys(filterVariables)).toHaveLength(0);
  });

  it("should not be empty object if params given", () => {
    const params: ProductListUrlFilters = {
      priceFrom: "10",
      priceTo: "20",
      status: true.toString(),
      stockStatus: StockAvailability.IN_STOCK
    };
    const filterVariables = getFilterVariables(params);

    expect(getExistingKeys(filterVariables)).toHaveLength(3);
  });
});

describe("Filtering URL params", () => {
  const intl = createIntl(config);

  const filters = createFilterStructure(intl, productListFilterOpts);

  it("should be empty if no active filters", () => {
    const filterQueryParams = getFilterQueryParams(
      filters,
      getFilterQueryParam
    );

    expect(getExistingKeys(filterQueryParams)).toHaveLength(0);
  });

  it("should not be empty if active filters are present", () => {
    const filterQueryParams = getFilterQueryParams(
      setFilterOptsStatus(filters, true),
      getFilterQueryParam
    );

    expect(filterQueryParams).toMatchSnapshot();
    expect(stringifyQs(filterQueryParams)).toMatchSnapshot();
  });
});
