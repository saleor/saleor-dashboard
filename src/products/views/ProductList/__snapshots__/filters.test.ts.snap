// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filtering URL params should not be empty if active filters are present 1`] = `
Object {
  "attributes": Object {
    "author": Array [
      "john-doe",
      false,
    ],
    "box-size": Array [
      "100g",
      "500g",
    ],
    "brand": Array [
      "saleor",
      false,
    ],
    "candy-box-size": Array [
      "100g",
      "500g",
    ],
    "coffee-genre": Array [
      "arabica",
      false,
    ],
    "collar": Array [
      "round",
      "polo",
    ],
    "color": Array [
      "blue",
      false,
    ],
    "cover": Array [
      "soft",
      "middle-soft",
    ],
    "flavor": Array [
      "sour",
      false,
    ],
    "language": Array [
      "english",
      false,
    ],
    "publisher": Array [
      "mirumee-press",
      false,
    ],
    "size": Array [
      "xs",
      "m",
    ],
  },
  "categories": Array [
    "878752",
  ],
  "collections": Array [
    "Q29sbGVjdGlvbjoc",
  ],
  "priceFrom": "10",
  "priceTo": "20",
  "productTypes": Array [
    "UHJvZHVjdFR5cGU6MQ==",
  ],
  "status": "published",
  "stockStatus": "IN_STOCK",
}
`;

exports[`Filtering URL params should not be empty if active filters are present 2`] = `"status=published&stockStatus=IN_STOCK&priceFrom=10&priceTo=20&categories%5B0%5D=878752&collections%5B0%5D=Q29sbGVjdGlvbjoc&productTypes%5B0%5D=UHJvZHVjdFR5cGU6MQ%3D%3D&attributes%5Bauthor%5D%5B0%5D=john-doe&attributes%5Bauthor%5D%5B1%5D=false&attributes%5Bbox-size%5D%5B0%5D=100g&attributes%5Bbox-size%5D%5B1%5D=500g&attributes%5Bbrand%5D%5B0%5D=saleor&attributes%5Bbrand%5D%5B1%5D=false&attributes%5Bcandy-box-size%5D%5B0%5D=100g&attributes%5Bcandy-box-size%5D%5B1%5D=500g&attributes%5Bcoffee-genre%5D%5B0%5D=arabica&attributes%5Bcoffee-genre%5D%5B1%5D=false&attributes%5Bcollar%5D%5B0%5D=round&attributes%5Bcollar%5D%5B1%5D=polo&attributes%5Bcolor%5D%5B0%5D=blue&attributes%5Bcolor%5D%5B1%5D=false&attributes%5Bcover%5D%5B0%5D=soft&attributes%5Bcover%5D%5B1%5D=middle-soft&attributes%5Bflavor%5D%5B0%5D=sour&attributes%5Bflavor%5D%5B1%5D=false&attributes%5Blanguage%5D%5B0%5D=english&attributes%5Blanguage%5D%5B1%5D=false&attributes%5Bpublisher%5D%5B0%5D=mirumee-press&attributes%5Bpublisher%5D%5B1%5D=false&attributes%5Bsize%5D%5B0%5D=xs&attributes%5Bsize%5D%5B1%5D=m"`;
